core/string/src/builder.rs

```diff
@@ -0,0 +1,423 @@
+        }
+    }
+
+    /// Returns the pointer of `data` of inner.
+    ///
+    /// # Safety
+    ///
+    /// Caller should ensure that the inner is allocated.
+    #[must_use]
+    unsafe fn data(&self) -> *mut D {
+        // SAFETY:
+        // Caller should ensure that the inner is allocated.
+        unsafe { addr_of_mut!((*self.inner.as_ptr()).data).cast() }
+    }
+
+    /// Inner logic of `allocate`.
+    ///
+    /// Use `realloc` here because it has a better performance than using `alloc`, `copy` and `dealloc`.
+    #[allow(clippy::cast_ptr_alignment)]
+    fn allocate_inner(&mut self, new_layout: Layout) {
+        let new_ptr = if self.is_allocated() {
+            let old_ptr = self.inner.as_ptr();
+            let old_layout = self.current_layout();
+            // SAFETY:
+            // Valid pointer is required by `realloc` and pointer is checked above to be valid.
+            // The layout size of `RawJsString` is never zero, since it has to store
+            // the length of the string and the reference count.
+            unsafe { realloc(old_ptr.cast(), old_layout, new_layout.size()) }
+        } else {
+            // SAFETY:
+            // The layout size of `RawJsString` is never zero, since it has to store
+            // the length of the string and the reference count.
+            unsafe { alloc(new_layout) }
+        };
+        let Some(new_ptr) = NonNull::new(new_ptr.cast::<RawJsString>()) else {
+            std::alloc::handle_alloc_error(new_layout)
+        };
+        self.inner = new_ptr;
+        self.cap = Self::capacity_from_layout(new_layout);
+    }
+
+    /// Appends an element to the inner of `JsStringBuilder`.
+    pub fn push(&mut self, v: D) {
+        let len = self.len();
+        if len == self.capacity() {
+            self.allocate(len + 1);
+        }
+        // SAFETY:
+        // Capacity has been expanded to be large enough to hold elements.
+        unsafe {
+            self.push_unchecked(v);
+        }
+    }
+
+    /// Push elements from slice to `JsStringBuilder` without doing capacity check.
+    ///
+    /// Unlike the standard vector, our holded element types are only `u8` and `u16`, which is [`Copy`] derived,
+    ///
+    /// so we only need to copy them instead of cloning.
+    ///
+    /// # Safety
+    ///
+    /// Caller should ensure the capacity is large enough to hold elements.
+    pub unsafe fn extend_from_slice_unchecked(&mut self, v: &[D]) {
+        // SAFETY: Caller should ensure the capacity is large enough to hold elements.
+        unsafe {
+            ptr::copy_nonoverlapping(v.as_ptr(), self.data().add(self.len()), v.len());
+        }
+        self.len += v.len();
+    }
+
+    /// push elements from slice to `JsStringBuilder`.
+    pub fn extend_from_slice(&mut self, v: &[D]) {
+        let required_cap = self.len() + v.len();
+        if required_cap > self.capacity() {
+            self.allocate(required_cap);
+        }
```

Comment:
Minor: Maybe we can factor out the check and call into a separate function something like `allocate_if_needed(cap: usize)`.